### PR TITLE
refactor: Migrate to Collections

### PR DESCRIPTION
**Pull request by jim380**
## Overview
This PR migrates the custom modules to Collections, a new module state management system introduced in SDK v0.50. This is a fundamental change to how we think of state in modules. 
- Before - state as raw bytes
- After - state as concrete golang objects

## Related
#116

### CODE REVIEW

**Comment by hacheigriega** [comment]
There is a slight chance this is relevant or helpful for the failing simulation tests.  I will address this PR in a few days.

**Review by DeshErBojhaa — APPROVED** [review]
utAck!

File: x/wasm-storage/keeper/keeper.go [hunk 1]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
+// SetProxyContractRegistry stores Proxy Contract address.
+func (k Keeper) SetProxyContractRegistry(ctx sdk.Context, address sdk.AccAddress) error {
+	return k.ProxyContractRegistry.Set(ctx, address)
+}
+
+// GetProxyContractRegistry returns Proxy Contract address.
+func (k Keeper) GetProxyContractRegistry(ctx sdk.Context) (sdk.AccAddress, error) {
+	return k.ProxyContractRegistry.Get(ctx)
+}
+
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:228]: We don't need these functions anymore.

File: x/wasm-storage/keeper/keeper.go [hunk 2]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
-	store := ctx.KVStore(k.storeKey)
-	bz, err := k.cdc.Marshal(&params)
-	if err != nil {
-		return err
-	}
-	store.Set([]byte(types.KeyParams), bz)
-
-	return nil
+	return k.Params.Set(ctx, params)
 }
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:241]: We don't need these methods anymore.

File: x/wasm-storage/keeper/keeper.go [hunk 3]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
 
 // HasOverlayWasm checks if a given Overlay Wasm exists.
-func (k Keeper) HasOverlayWasm(ctx sdk.Context, wasm *types.Wasm) bool {
-	store := ctx.KVStore(k.storeKey)
-	return store.Has(types.GetOverlayWasmKey(wasm.Hash))
+func (k Keeper) HasOverlayWasm(ctx sdk.Context, wasm *types.Wasm) (bool, error) {
+	overlayWasmKeyPrefixFull := GetOverlayWasmKeyPrefixFull(wasm.Hash)
+
+	return k.OverlayWasm.Has(ctx, overlayWasmKeyPrefixFull)
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:126]: Can be merged into single line of code

File: x/wasm-storage/keeper/keeper.go [hunk 4]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
 // GetOverlayWasm returns Overlay Wasm given its key.
-func (k Keeper) GetOverlayWasm(ctx sdk.Context, hash []byte) *types.Wasm {
-	var wasm types.Wasm
-	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.GetOverlayWasmKey(hash))
-	k.cdc.MustUnmarshal(bz, &wasm)
-	return &wasm
+func (k Keeper) GetOverlayWasm(ctx sdk.Context, hash []byte) (*types.Wasm, error) {
+	overlayWasmKeyPrefixFull := GetOverlayWasmKeyPrefixFull(hash)
+	wasm, err := k.OverlayWasm.Get(ctx, overlayWasmKeyPrefixFull)
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:114]: Can be merged into single line without declaring a variable.

File: x/wasm-storage/keeper/keeper.go [hunk 5]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
 }
 
 // SetOverlayWasm stores Overlay Wasm using its hash as the key.
-func (k Keeper) SetOverlayWasm(ctx sdk.Context, wasm *types.Wasm) {
-	store := ctx.KVStore(k.storeKey)
-	bz := k.cdc.MustMarshal(wasm)
-	store.Set(types.GetOverlayWasmKey(wasm.Hash), bz)
+func (k Keeper) SetOverlayWasm(ctx sdk.Context, wasm *types.Wasm) error {
+	overlayWasmKeyPrefixFull := GetOverlayWasmKeyPrefixFull(wasm.Hash)
+	return k.OverlayWasm.Set(ctx, overlayWasmKeyPrefixFull, *wasm)
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:108]: Can be merged into single line

File: x/wasm-storage/keeper/keeper.go [hunk 6]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
 }
 
 // HasDataRequestWasm checks if a given Data Request Wasm exists.
-func (k Keeper) HasDataRequestWasm(ctx sdk.Context, wasm *types.Wasm) bool {
-	store := ctx.KVStore(k.storeKey)
-	return store.Has(types.GetDataRequestWasmKey(wasm.Hash))
+func (k Keeper) HasDataRequestWasm(ctx sdk.Context, wasm *types.Wasm) (bool, error) {
+	dataRequestWasmKeyPrefixFull := GetDataRequestWasmKeyPrefixFull(wasm.Hash)
+
+	return k.DataRequestWasm.Has(ctx, dataRequestWasmKeyPrefixFull)
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:102]: Can be merged into single line

File: x/wasm-storage/keeper/keeper.go [hunk 7]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
 // GetDataRequestWasm returns Data Request Wasm given its key.
-func (k Keeper) GetDataRequestWasm(ctx sdk.Context, hash []byte) *types.Wasm {
-	var wasm types.Wasm
-	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.GetDataRequestWasmKey(hash))
-	k.cdc.MustUnmarshal(bz, &wasm)
-	return &wasm
+func (k Keeper) GetDataRequestWasm(ctx sdk.Context, hash []byte) (*types.Wasm, error) {
+	dataRequestWasmKeyPrefixFull := GetDataRequestWasmKeyPrefixFull(hash)
+	wasm, err := k.DataRequestWasm.Get(ctx, dataRequestWasmKeyPrefixFull)
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:90]: can be merged into single line without declaring a variable

File: x/wasm-storage/keeper/keeper.go [hunk 8]
```diff
@@ -35,139 +78,164 @@ func (k Keeper) GetAuthority() string {
...
 
 // SetDataRequestWasm stores Data Request Wasm using its hash as the key.
-func (k Keeper) SetDataRequestWasm(ctx sdk.Context, wasm *types.Wasm) {
-	store := ctx.KVStore(k.storeKey)
-	bz := k.cdc.MustMarshal(wasm)
-	store.Set(types.GetDataRequestWasmKey(wasm.Hash), bz)
+func (k Keeper) SetDataRequestWasm(ctx sdk.Context, wasm *types.Wasm) error {
+	dataRequestWasmKeyPrefixFull := GetDataRequestWasmKeyPrefixFull(wasm.Hash)
+
+	return k.DataRequestWasm.Set(ctx, dataRequestWasmKeyPrefixFull, *wasm)
```
Comment (hacheigriega) [inline x/wasm-storage/keeper/keeper.go:84]: can be merged into single line without declaring a variable

**Review by hacheigriega — CHANGES_REQUESTED** [review]

File: x/wasm-storage/keeper/keeper.go (hunk 1 above)
Comment (DeshErBojhaa) [reply x/wasm-storage/keeper/keeper.go]: Refactored here [3310202](https://github.com/sedaprotocol/seda-chain/pull/192/commits/3310202f76cc5b82224800e2233e934a90f589f1)

File: x/wasm-storage/keeper/keeper.go (hunk 2 above)
Comment (DeshErBojhaa) [reply x/wasm-storage/keeper/keeper.go]: Refactor: removed func [e1aaee3](https://github.com/sedaprotocol/seda-chain/pull/192/commits/e1aaee3867fdf3b44333219846a375dda0bb8c6c)

File: x/wasm-storage/keeper/keeper.go (hunk 3 above)
Comment (DeshErBojhaa) [reply x/wasm-storage/keeper/keeper.go]: [62cfbab](https://github.com/sedaprotocol/seda-chain/pull/192/commits/62cfbabec5a4ad9788cf1ae06c725cf9b2b39a4b)

File: x/wasm-storage/keeper/keeper.go (hunk 5 above)
Comment (DeshErBojhaa) [reply x/wasm-storage/keeper/keeper.go]: [62cfbab](https://github.com/sedaprotocol/seda-chain/pull/192/commits/62cfbabec5a4ad9788cf1ae06c725cf9b2b39a4b)

File: x/wasm-storage/keeper/keeper.go (hunk 6 above)
Comment (DeshErBojhaa) [reply x/wasm-storage/keeper/keeper.go]: [62cfbab](https://github.com/sedaprotocol/seda-chain/pull/192/commits/62cfbabec5a4ad9788cf1ae06c725cf9b2b39a4b)

File: x/wasm-storage/keeper/keeper.go (hunk 7 above)
Comment (DeshErBojhaa) [reply x/wasm-storage/keeper/keeper.go]: [62cfbab](https://github.com/sedaprotocol/seda-chain/pull/192/commits/62cfbabec5a4ad9788cf1ae06c725cf9b2b39a4b)

File: x/wasm-storage/keeper/keeper.go (hunk 8 above)
Comment (DeshErBojhaa) [reply x/wasm-storage/keeper/keeper.go]: [62cfbab](https://github.com/sedaprotocol/seda-chain/pull/192/commits/62cfbabec5a4ad9788cf1ae06c725cf9b2b39a4b)

**Pull request merged**